From ff0bc084ebde676504954c58be58e2511a866632 Mon Sep 17 00:00:00 2001
From: Yangbo Lu <yangbo.lu@nxp.com>
Date: Tue, 31 Jul 2018 11:15:06 +0800
Subject: [PATCH 182/706] enetc: add get_ts_info interface for ethtool

This patch is to add get_ts_info interface for ethtool
to support getting timestamping capability.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>
Signed-off-by: Claudiu Manoil <claudiu.manoil@nxp.com>
(cherry picked from commit 00f403a193878d99b349887548f512d5c97dda1a)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/net/ethernet/freescale/enetc/enetc.h  |  4 +++
 .../ethernet/freescale/enetc/enetc_ethtool.c  | 33 +++++++++++++++++++
 .../net/ethernet/freescale/enetc/enetc_ptp.c  |  7 ++--
 3 files changed, 41 insertions(+), 3 deletions(-)

diff --git a/drivers/net/ethernet/freescale/enetc/enetc.h b/drivers/net/ethernet/freescale/enetc/enetc.h
index 032638c9b511..5898841f4e70 100644
--- a/drivers/net/ethernet/freescale/enetc/enetc.h
+++ b/drivers/net/ethernet/freescale/enetc/enetc.h
@@ -265,5 +265,9 @@ void enetc_tsn_init(struct enetc_si *si);
 #define enetc_tsn_init(si) (void)0
 #endif
 
+/* PTP driver exports */
+#define ENETC_PHC_INDEX_DEFAULT	-1
+extern int enetc_phc_index;
+
 /* common PF and VF module params */
 extern unsigned int debug;
diff --git a/drivers/net/ethernet/freescale/enetc/enetc_ethtool.c b/drivers/net/ethernet/freescale/enetc/enetc_ethtool.c
index 50319c809f82..93a60d1f3fe8 100644
--- a/drivers/net/ethernet/freescale/enetc/enetc_ethtool.c
+++ b/drivers/net/ethernet/freescale/enetc/enetc_ethtool.c
@@ -32,6 +32,8 @@
  * POSSIBILITY OF SUCH DAMAGE.
  */
 
+#include <linux/net_tstamp.h>
+#include <linux/module.h>
 #include "enetc.h"
 
 static const u32 enetc_si_regs[] = {
@@ -572,6 +574,35 @@ static void enetc_get_ringparam(struct net_device *ndev,
 	}
 }
 
+static int enetc_get_ts_info(struct net_device *ndev,
+			     struct ethtool_ts_info *info)
+{
+	int *phc_idx;
+
+	phc_idx = symbol_get(enetc_phc_index);
+	if (phc_idx) {
+		info->phc_index = *phc_idx;
+		symbol_put(enetc_phc_index);
+	} else {
+		info->phc_index = ENETC_PHC_INDEX_DEFAULT;
+	}
+
+#ifdef CONFIG_ENETC_HW_TIMESTAMPING
+	info->so_timestamping = SOF_TIMESTAMPING_TX_HARDWARE |
+				SOF_TIMESTAMPING_RX_HARDWARE |
+				SOF_TIMESTAMPING_RAW_HARDWARE;
+
+	info->tx_types = (1 << HWTSTAMP_TX_OFF) |
+			 (1 << HWTSTAMP_TX_ON);
+	info->rx_filters = (1 << HWTSTAMP_FILTER_NONE) |
+			   (1 << HWTSTAMP_FILTER_ALL);
+#else
+	info->so_timestamping = SOF_TIMESTAMPING_RX_SOFTWARE |
+				SOF_TIMESTAMPING_SOFTWARE;
+#endif
+	return 0;
+}
+
 const struct ethtool_ops enetc_pf_ethtool_ops = {
 	.get_regs_len = enetc_get_reglen,
 	.get_regs = enetc_get_regs,
@@ -585,6 +616,7 @@ const struct ethtool_ops enetc_pf_ethtool_ops = {
 	.get_rxfh = enetc_get_rxfh,
 	.set_rxfh = enetc_set_rxfh,
 	.get_ringparam = enetc_get_ringparam,
+	.get_ts_info = enetc_get_ts_info,
 };
 
 const struct ethtool_ops enetc_vf_ethtool_ops = {
@@ -599,6 +631,7 @@ const struct ethtool_ops enetc_vf_ethtool_ops = {
 	.get_rxfh = enetc_get_rxfh,
 	.set_rxfh = enetc_set_rxfh,
 	.get_ringparam = enetc_get_ringparam,
+	.get_ts_info = enetc_get_ts_info,
 };
 
 void enetc_set_ethtool_ops(struct net_device *ndev)
diff --git a/drivers/net/ethernet/freescale/enetc/enetc_ptp.c b/drivers/net/ethernet/freescale/enetc/enetc_ptp.c
index 3054bc05e6fc..4ed3bed358d0 100644
--- a/drivers/net/ethernet/freescale/enetc/enetc_ptp.c
+++ b/drivers/net/ethernet/freescale/enetc/enetc_ptp.c
@@ -6,6 +6,9 @@
 
 #include "enetc_ptp.h"
 
+int enetc_phc_index = ENETC_PHC_INDEX_DEFAULT;
+EXPORT_SYMBOL(enetc_phc_index);
+
 struct enetc_ptp_config {
 	u32	ck_sel;
 	u32	tclk_period;
@@ -24,8 +27,6 @@ struct enetc_ptp {
 	spinlock_t lock; /* protects regs */
 };
 
-int enetc_phc_index;
-
 static u64 tmr_cnt_read(struct enetc_ptp *ptp)
 {
 	u64 ns;
@@ -340,7 +341,7 @@ static void enetc_ptp_remove(struct pci_dev *pdev)
 	enetc_wr_reg(ptp->regs + TMR_TEVENT, 0);
 	enetc_wr_reg(ptp->regs + TMR_CTRL, 0);
 
-	enetc_phc_index = -1;
+	enetc_phc_index = ENETC_PHC_INDEX_DEFAULT;
 	ptp_clock_unregister(ptp->clock);
 
 	iounmap(ptp->regs);
-- 
2.17.1

